x-pack/platform/plugins/shared/onechat/server/routes/mcp.ts

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
+import { schema } from '@kbn/config-schema';
+import type { RouteDependencies } from './types';
+import { getHandlerWrapper } from './wrap_handler';
+import { KibanaMcpHttpTransport } from '../utils/kibana_mcp_http_transport';
+
+const TECHNICAL_PREVIEW_WARNING =
+  'Elastic MCP Server is in technical preview and may be changed or removed in a future release. Elastic will work to fix any issues, but features in technical preview are not subject to the support SLA of official GA features.';
+
+const MCP_SERVER_NAME = 'elastic-mcp-server';
+const MCP_SERVER_VERSION = '0.0.1';
+const MCP_SERVER_PATH = '/api/onechat/mcp';
+
+export function registerMCPRoutes({ router, getInternalServices, logger }: RouteDependencies) {
+  const wrapHandler = getHandlerWrapper({ logger });
+
+  router.versioned
+    .post({
+      path: MCP_SERVER_PATH,
+      security: {
+        authz: {
+          enabled: false,
+          reason: 'todo',
+        },
+      },
+      access: 'public',
+      summary: 'MCP server',
+      description: TECHNICAL_PREVIEW_WARNING,
+      options: {
+        tags: ['mcp'],
+        xsrfRequired: false,
+        availability: {
+          stability: 'experimental',
+        },
+      },
+    })
+    .addVersion(
+      {
+        version: '2023-10-31',
+        validate: {
+          request: { body: schema.object({}, { unknowns: 'allow' }) },
+        },
+      },
+      wrapHandler(async (_, request, response) => {
+        let transport: KibanaMcpHttpTransport | undefined;
+        let server: McpServer | undefined;
+
+        try {
+          transport = new KibanaMcpHttpTransport({ sessionIdGenerator: undefined, logger });
+
+          // Instantiate new MCP server upon every request, no session persistence
+          server = new McpServer({
```

Comment:
Instantiating a new transport and server on each request feels odd to me. Should these not be reused across requests?